velox/docs/functions/spark/aggregate.rst

```diff
@@ -11,6 +11,22 @@ General Aggregate Functions
 
     Returns the bitwise XOR of all non-null input values, or null if none.
 
+.. spark:function:: bloom_filter_agg(x, estimatedNumItems, numBits) -> varbinary
+
+    Insert ``x`` into BloomFilter, and returns the serialized BloomFilter.
```

Comment:
nit: perhaps,

```
Creates bloom filter from values of 'x' and returns it serialized into VARBINARY.

``estimatedNumItems`` provides an estimate of the number of unique values of ``x``. Value is capped at 716,800.

``numBits`` specifies max capacity of the bloom filter, which allows to trade accuracy for memory.
```

Reply:
Change `the number of unique values` to `the number of values`, this is spark intend meaning.

Reply:
@jinchengchenghh I'm not sure I understand why would we want to specify the estimate of the total number of input values. Would you clarify?

Reply:
```
new BloomFilterImpl(optimalNumOfHashFunctions(expectedNumItems, numBits), numBits);
  /**
   * Computes the optimal k (number of hashes per item inserted in Bloom filter), given the
   * expected insertions and total number of bits in the Bloom filter.
   *
   * See http://en.wikipedia.org/wiki/File:Bloom_filter_fp_probability.svg for the formula.
   *
   * @param n expected insertions (must be positive)
   * @param m total number of bits in Bloom filter (must be positive)
   */
  private static int optimalNumOfHashFunctions(long n, long m) {
    // (m / n) * log(2), but avoid truncation due to division!
    return Math.max(1, (int) Math.round((double) m / n * Math.log(2)));
  }
```

The spark BloomFilter implementation use this value to compute number of hash functions, it has a optimal value according to the theory. Velox implementation does not use variable `number of hash function`, it uses constant 4

Reply:
@jinchengchenghh Are you saying Velox's implementation doesn't use estimatedNumItems arguments? If so, should we remove it? Otherwise, should we document that this argument is not used and remove logic for capping its value and initializing estimatedNumItems_ member variable.

Reply:
For the final solution, I may implement native Spark BloomFilter in Velox, it will have better performance, we can switch to it then.

And if we not specify `numBits`, we will use `estimatedNumItems` to estimate `numBits`, BloomFilter implemenation doesn't use estimatedNumItems argument, but used in bloom_filter_agg

Reply:
Got it. Let's clarify all this in the documentation. It is not obvious.

Reply:
Do we need to also document the difference between spark, or just clarify the usage in Velox?

Reply:
Let's document the difference with Spark as well. By default, the assumption is that Velox functions match semantics of the original engine.

Reply:
Ok, updated